cpp/src/spatial/polygon_distance.cu

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required point_b_y applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuspatial/detail/cartesian_product_group_index_iterator.cuh>
+#include <cuspatial/error.hpp>
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/iterator/discard_iterator.h>
+
+#include <limits>
+#include <memory>
+
+using size_type = cudf::size_type;
+
+namespace cuspatial {
+namespace detail {
+namespace {
+
+/**
+ * @brief Calculates segment-point distance
+ *
+ * Given a `cartesian_product_group_index` and two columns representing `x` and `y` coordinates,
+ * calculates the segment-point distance of a line segment in group `a` and a point in group `b`.
+ *
+ * If group `a` contains two or more points, calculate segment-point distance.
+ * If group `a` contains only a single point, calculate point-point distance.
+ */
+template <typename T>
+struct segment_point_distance_calculator {
+  cudf::column_device_view xs;
+  cudf::column_device_view ys;
+
+  T __device__ operator()(cartesian_product_group_index idx)
+  {
+    auto const a_idx_0 = idx.group_a.offset + (idx.element_a_idx);
+    auto const a_idx_1 = idx.group_a.offset + (idx.element_a_idx + 1) % idx.group_a.size;
+    auto const b_idx_0 = idx.group_b.offset + (idx.element_b_idx);
+
+    auto const origin_x = xs.element<T>(a_idx_0);
+    auto const origin_y = ys.element<T>(a_idx_0);
+    auto const edge_x   = xs.element<T>(a_idx_1) - origin_x;
+    auto const edge_y   = ys.element<T>(a_idx_1) - origin_y;
+    auto const point_x  = xs.element<T>(b_idx_0) - origin_x;
+    auto const point_y  = ys.element<T>(b_idx_0) - origin_y;
+
+    auto const magnitude = edge_x * edge_x + edge_y * edge_y;
+    auto const travel    = point_x * edge_x + point_y * edge_y;
+
+    // if point is projected within line segment bounds, use segment-point distance.
+    // if point is projected outside line segment bounds, use point-point distance.
+    if (0 < travel && travel < magnitude) {  // 0 < travel < edge_length
```

Comment:
```suggestion
    if (0 < travel && travel < magnitude) {
```

Reply:
Hah. I was originally using `travel * (magnitude - travel) >= 0`, which is why the comment was there. Thanks! :laughing: